### ISSUE TITLE
`thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Other("Is a directory (os error 21)")', lib/src/operation.rs:97:64`

### ISSUE THREAD

**Issue by ocharles** [jj-vcs/jj#1907]
## Description

My machine locked up and needed a hard reset, so there's a chance there's some file system corruption here (I run ext4). When I run any `jj` command, I now get:

```
❯ RUST_BACKTRACE=full jj op log
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Other("Is a directory (os error 21)")', lib/src/operation.rs:97:64
stack backtrace:
   0:     0x55e0419ba4f1 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hded4932df41373b3
   1:     0x55e0419e492f - core::fmt::write::hb1cb56105a082ad9
   2:     0x55e0419b6091 - std::io::Write::write_fmt::h797fda7085c97e57
   3:     0x55e0419ba305 - std::sys_common::backtrace::print::hf74aa2eef05af215
   4:     0x55e0419bb9d7 - std::panicking::default_hook::{{closure}}::h8cad394227ea3de8
   5:     0x55e0419bb7c4 - std::panicking::default_hook::h249cc184fec99a8a
   6:     0x55e0419bbe8c - std::panicking::rust_panic_with_hook::h82ebcd5d5ed2fad4
   7:     0x55e0419bbd87 - std::panicking::begin_panic_handler::{{closure}}::h810bed8ecbe66f1a
   8:     0x55e0419ba926 - std::sys_common::backtrace::__rust_end_short_backtrace::h1410008071796261
   9:     0x55e0419bbad2 - rust_begin_unwind
  10:     0x55e0413e4f23 - core::panicking::panic_fmt::ha0a42a25e0cf258d
  11:     0x55e0413e53c3 - core::result::unwrap_failed::h100c4d67576990cf
  12:     0x55e04178213b - jj_lib::operation::Operation::view::h6709b46a87292520
  13:     0x55e041788c0b - jj_lib::repo::RepoLoader::load_at::h94367482d5849691
  14:     0x55e04147a6d4 - jj_cli::cli_util::CommandHelper::workspace_helper_internal::h2cdc6b2db1464e83
  15:     0x55e0414b8139 - jj_cli::commands::operation::cmd_op_log::h010be4d923fb200e
  16:     0x55e0414d90df - jj_cli::commands::run_command::h44a341f058127919
  17:     0x55e041418880 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h8fe6a69f8c1091b8
  18:     0x55e04146bf55 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h0847077649ffbc54
  19:     0x55e04149d1e8 - jj_cli::cli_util::CliRunner::run::h0dba531a0a7d9e9c
  20:     0x55e0413e57d1 - jj::main::h4e8ca46c819e8247
  21:     0x55e0413e5753 - std::sys_common::backtrace::__rust_begin_short_backtrace::h59ca8584147632a6
  22:     0x55e0413e5769 - std::rt::lang_start::{{closure}}::hd1da4e9f7a4e8af7
  23:     0x55e0419ae795 - std::rt::lang_start_internal::hd66bf6b7da144005
  24:     0x55e0413e5805 - main
  25:     0x7f60be55dace - __libc_start_call_main
  26:     0x7f60be55db89 - __libc_start_main@@GLIBC_2.34
  27:     0x55e0413e5685 - _start
  28:                0x0 - <unknown>
```

Not sure where to go from here, but I've archived this repo if we want to do any further investigations!

**Comment by ocharles** [comment]
Hmm, I also ran out of space while making a backup of the repository, so I might have also ran out of disk space when the machine locked up. #thisisfine :sweat:

**Comment by martinvonz** [comment]
Weird. It sounds like a something (a view object) that's normally a file has become a directory. Do you have any directories in ` .jj/repo/op_store/views/`? It should look like this:
```
$ find .jj/repo/op_store/ -type d
.jj/repo/op_store/
.jj/repo/op_store/views
.jj/repo/op_store/operations
```

Are you able to share the archived repo?

**Comment by ocharles** [comment]
Unfortunately I can't share the repo as it's the entire mono-repo for work. I tried that command and got the same:

```
❯ find .jj/repo/op_store/ -type d
.jj/repo/op_store/
.jj/repo/op_store/operations
.jj/repo/op_store/views
```

Given how generally fucked this desktop is atm (constant lockups leading to hard reboots and almost definite data corruption) it's likely that this is just a weird file somewhere that's in an odd state, so I'm happy to close as wontfix if there's not enough reproducible info. I thought I'd raise it just in case we can help generate a better error message.

**Comment by martinvonz** [comment]
Yes, we can definitely improve that error message and improve error reporting in general, so feel free to leave this open for tracking that.

**Comment by martinvonz** [comment]
@ocharles, with #1908 merged, do you want to see if you get a more useful error message?

**Comment by ocharles** [comment]
A bit better!

```
❯ jj
Hint: Use `jj -h` for a list of available commands.
Set the config `ui.default-command = "log"` to disable this message.
Internal error: Failed to load the repo: Is a directory (os error 21)
```

It would be nice to what unexpectedly a directory, though.

**Comment by martinvonz** [comment]
> It would be nice to what unexpectedly a directory, though.

How about now (after #1915)? :)

**Comment by ocharles** [comment]
Hmm, not sure that one made a difference :thinking: 

```
circuithub-jj-1907 (d892795) via λ 
❯ jj --version
jj 0.8.0-4834d12c37e5863c6db453c5553e7b0d4c8827cf

circuithub-jj-1907 (d892795) via λ 
❯ jj
Hint: Use `jj -h` for a list of available commands.
Set the config `ui.default-command = "log"` to disable this message.
Internal error: Failed to load the repo: Error when reading object  of type view: Is a directory (os error 21)
```

**Comment by ocharles** [comment]
> Can you also run cat .jj/repo/op_store/operations/$(basename $(ls .jj/repo/op_heads/heads/*)) and let me know what that says? 

```
❯ bat .jj/repo/op_store/operations/(basename (ls .jj/repo/op_heads/heads/*))
───────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: .jj/repo/op_store/operations/18fa67b2c31fe7a3bc15608039eae587d8e57c95b6eb569d10d956fa1155cb3c83a9053af3b10a09b8eca133fc8f79c89defed53a372b456e2457e3a1a245025   <EMPTY>
───────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Comment by martinvonz** [comment]
Yeah, that's what I expected. I don't know how it could have happened, however. The code for writing an operation is here:
https://github.com/martinvonz/jj/blob/4834d12c37e5863c6db453c5553e7b0d4c8827cf/lib/src/simple_op_store.rs#L134-L148

It's changed a bit since your bug report, but it didn't ignore errors from either the file-write or the rename even before. And the rename is atomic. So the only thing I can think is that it failed to flush to disk after we thought the write had succeeded. Maybe some `fsync()` calls are required to make this safe.

**Comment by ocharles** [comment]
I'm gonna close this now as I don't think I even have the repro repo around any more.
